src/main/java/hudson/tasks/MailSender.java

```diff
@@ -401,66 +377,80 @@ private MimeMessage createEmptyMail(Run<?, ?> build, TaskListener listener) thro
             if(debug)
                 listener.getLogger().println("Trying to send e-mails to individuals who broke the build. sizeof(culprits)=="+culprits.size());
 
-            rcp.addAll(buildCulpritList(listener,culprits));
+            messageBuilder.addRecipients(getUserEmailList(listener, culprits));
         }
         
         // set recipients after filtering out recipients that should not receive emails
-        Set<InternetAddress> filteredRecipients = MailAddressFilter.filterRecipients(build, listener, rcp);
-        if (!filteredRecipients.isEmpty()) {
-            msg.setRecipients(Message.RecipientType.TO,
-                    filteredRecipients.toArray(new InternetAddress[filteredRecipients.size()]));
-        }
+        messageBuilder.setRecipientFilter(new MimeMessageBuilder.AddressFilter() {
+            @Override
+            public Set<InternetAddress> apply(Set<InternetAddress> recipients) {
+                return MailAddressFilter.filterRecipients(build, listener, recipients);
+            }
+        });
```

Comment:
Hmm, if you need to do this then hiding `toNormalizedAddress` is making the API more complicated, not less.